test/functional/runner/index.js

```diff
@@ -0,0 +1,590 @@
+        {
+          command: command.command,
+          commandName: command.command_name ? command.command_name : command.commandName,
+          databaseName: command.database_name ? command.database_name : command.databaseName
+        },
+        { relaxed: true }
+      )
+    )
+  );
+}
+
+function extractCrudResult(result, operation) {
+  if (Array.isArray(result) || !isPlainObject(result)) {
+    return result;
+  }
+
+  if (result.value) {
+    // some of our findAndModify results return more than just an id, so we need to pluck
+    const resultKeys = Object.keys(operation.result);
+    if (resultKeys.length === 1 && resultKeys[0] === '_id') {
+      return { _id: result.value._id };
+    }
+
+    return result.value;
+  }
+
+  return Object.keys(operation.result).reduce((crudResult, key) => {
+    if (result.hasOwnProperty(key) && result[key] != null) {
+      // FIXME(major): update crud results are broken and need to be changed
+      crudResult[key] = key === 'upsertedId' ? result[key]._id : result[key];
+    }
+
+    return crudResult;
+  }, {});
+}
+
+function isTransactionCommand(command) {
+  return ['startTransaction', 'commitTransaction', 'abortTransaction'].indexOf(command) !== -1;
+}
+
+function isTestRunnerCommand(command) {
+  return (
+    ['targetedFailPoint', 'assertSessionPinned', 'assertSessionUnpinned'].indexOf(command) !== -1
+  );
+}
+
+function extractBulkRequests(requests) {
+  return requests.map(request => ({ [request.name]: request.arguments }));
+}
+
+function translateOperationName(operationName) {
+  if (operationName === 'runCommand') return 'command';
+  return operationName;
+}
+
+function normalizeReadPreference(mode) {
+  return mode.charAt(0).toLowerCase() + mode.substr(1);
+}
+
+/**
+ *
+ * @param {Object} operation the operation definition from the spec test
+ * @param {Object} obj the object to call the operation on
+ * @param {Object} context a context object containing sessions used for the test
+ * @param {Object} [options] Optional settings
+ * @param {Boolean} [options.swallowOperationErrors] Generally we want to observe operation errors, validate them against our expectations, and then swallow them. In cases like `withTransaction` we want to use the same `testOperations` to build the lambda, and in those cases it is not desireable to swallow the errors, since we need to test this behavior.
+ */
+function testOperation(operation, obj, context, options) {
+  options = options || { swallowOperationErrors: true };
+  const opOptions = {};
+  let args = [];
+  const operationName = translateOperationName(operation.name);
+
+  if (operation.arguments) {
+    Object.keys(operation.arguments).forEach(key => {
+      if (key === 'callback') {
+        args.push(() =>
+          testOperations(operation.arguments.callback, context, { swallowOperationErrors: false })
+        );
+        return;
+      }
+
+      if (['filter', 'fieldName', 'document', 'documents', 'pipeline'].indexOf(key) !== -1) {
```

Comment:
Probably good to stick these in a set?

Reply:
I actually think this section of the runner is ripe for refactoring. There's a lot of weird stuff here around positional parameters (some of these branches `unshift`, others `shift`, others still add to `opOptions`.. I think we need a mapping of "models" for parameters to given methods. I'm going to hold off on using a `Set` here because I'm not even sure what I'd name the constant, let's rethink in follow-on work?